onnxruntime/core/providers/trt/trt_execution_provider.cc

```diff
@@ -0,0 +1,420 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "trt_execution_provider.h"
+#include "trt_allocator.h"
+#include "core/framework/execution_provider.h"
+#include "core/framework/op_kernel.h"
+#include "core/framework/kernel_registry.h"
+#include "core/framework/compute_capability.h"
+#include "core/providers/cpu/cpu_execution_provider.h"
+#include "core/platform/env.h"
+#include "onnx/shape_inference/implementation.h"
+#include "cuda_runtime_api.h"
+#include "gsl/pointers"
+#include "core/graph/model.h"
+#include "cuda_runtime_api.h"
+
+using namespace std;
+using namespace ONNX_NAMESPACE;
+using namespace ::onnxruntime::logging;
+
+namespace onnxruntime {
+
+TRTExecutionProvider::TRTExecutionProvider()
+    : IExecutionProvider{onnxruntime::kTRTExecutionProvider} {
+  DeviceAllocatorRegistrationInfo trt_device_info({OrtMemTypeCPU, [](int) {
+                                                     return std::make_unique<TRTPinnedAllocator>();
+                                                   },
+                                                   std::numeric_limits<size_t>::max()});
+  InsertAllocator(CreateAllocator(trt_device_info));
+  DeviceAllocatorRegistrationInfo default_device_info({OrtMemTypeDefault, [](int) {
+                                                         return std::make_unique<TRTAllocator>();
+                                                       },
+                                                       std::numeric_limits<size_t>::max()});
+  InsertAllocator(CreateAllocator(default_device_info));
+}
+
+TRTExecutionProvider::~TRTExecutionProvider() {}
+
+std::vector<std::unique_ptr<ComputeCapability>>
+TRTExecutionProvider::GetCapability(const onnxruntime::GraphViewer& graph,
+                                    const std::vector<const KernelRegistry*>& /*kernel_registries*/) const {
+  // Construct modelproto from graph
+  onnxruntime::Model model(graph.Name(), true, ModelMetaData(), IOnnxRuntimeOpSchemaRegistryList(), graph.DomainToVersionMap());
+  onnxruntime::Graph& graph_build = model.MainGraph();
+  const std::vector<NodeIndex>& node_index = graph.GetNodesInTopologicalOrder();
+  for (auto& index : node_index) {
+    const Node* node = graph.GetNode(index);
+    graph_build.AddNode(*node);
+  }
+  ORT_ENFORCE(graph_build.Resolve().IsOK());
+  ONNX_NAMESPACE::ModelProto model_proto = model.ToProto();
+  model_proto.set_ir_version(ONNX_NAMESPACE::Version::IR_VERSION);
+
+  // Serialize modelproto to string
+  string string_buf;
+  model_proto.SerializeToString(&string_buf);
+  std::vector<char> onnx_buf(string_buf.begin(), string_buf.end());
```

Comment:
Why do you need to create onnx_buf? This will create a copy and will be inefficient. Does passing string_buf.data() to supportsModel() not work?